api/types/entrySchema.go

```diff
@@ -0,0 +1,181 @@
+package apitypes
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/emirpasic/gods/maps/linkedhashmap"
+	"github.com/getlantern/deepcopy"
+	"github.com/puppetlabs/wash/plugin"
+)
+
+// EntrySchema describes an entry's schema, which
+// is what's returned by the /fs/schema endpoint.
+type EntrySchema struct {
+	plugin.EntrySchema
+	children map[string]*EntrySchema
+	// graph is an ordered map of <TypeID> => <EntrySchema>. We store it to make
+	// MarshalJSON's implementation easier.
+	//
+	// NOTE: The reason we don't synchronize children with graph is b/c entry
+	// schemas are immutable. Clients that want to mess with the state (e.g. like
+	// `wash find`) can do so at their own peril.
+	graph *linkedhashmap.Map
```

Comment:
linkedhashmap marshals/unmarshals/iterates in insertion order, which is why I'm using it here.

Reply:
If they need to be ordered, why not serialize as an array of objects? We can reconstruct the map from that easily .enough.

Reply:
I don't like the array because it doesn't structurally enforce the "TypeIDs are unique" requirement of entry schemas, which is fundamental. It's also true that maps don't have to enforce insertion order either, but it is also not as uncommon to consider an insertion order when you're reasoning about map keys vs. considering uniqueness when you're reasoning about arrays. For example, I can't help but think about insertion order when I look at JSON objects, while when I think about array uniqueness, I tend to think about things as sets rather than arrays.

Also, other languages have very easy-to-use ordered map structures. For example, Ruby hashes already let you iterate in insertion order. Python has OrderedDict which is also very easy to use. Even Java has the LinkedHashMap class. It's just Go that sucks (mostly due to its lack of generics support), so if it is only one language that happens to make enforcing insertion order suck, and that level of suck is not too bad (which I don't think it is here), then I would prefer returning a response that's more consistent with how we think about schemas. In this case, that response is a JSON object with the extra caveat that the first key must be the invoked entry's schema.